rfcs/20251203-grouped-gemm-support/README.md

```diff
@@ -0,0 +1,604 @@
+# Add Support for Grouped GEMM for Mixture of Experts (MoE)
+
+## Definition of the Operation
+
+Grouped GEMM is a computational kernel used in Mixture of Experts (MoE).
+
+Typical MoE workflow:
+- Input tokens are analyzed by a router
+- Router assigns each token to top-K experts based on learned scores (*)
+- Tokens are regrouped by their assigned experts (**)
+- Each expert processes its assigned tokens independently (this is where Grouped GEMM appears)
+- Expert outputs are combined back into the original token order
+
+(*) Routing creates variable-sized batches of tokens per expert. Some experts may
+receive many tokens while others receive few or none (distribution could be expected to be very unbalanced).
+The number of tokens per expert (`M` dimension) is only known after routing completes.
+
+(**) Regrouping involves gathering all tokens assigned to each expert
+and creating a contiguous block of memory for each expert's input.
+
+Grouped GEMM processes variable-sized batches (subsets of tokens) per expert,
+which is a key difference from batched GEMM operations that assume uniform batch sizes across groups.
+
+For each expert `i` (dimensions are: `src[Mi x K]`, `weights[K x N]`, `dst[Mi x N]`) we need to compute:
```

Comment:
The inner dimension is not fixed in the backward pass during training.

Consider,

$\mathbf{d\mathbf{W}_i} = \mathbf{S}_i^T \cdot \mathbf{d\mathbf{O}_i}$

With dimensions,

$\[\mathbf{K} \times \mathbf{N}\] = \[ \mathbf{K} \times \mathbf{M}_i \] \cdot \[ \mathbf{M}_i \times \mathbf{N}\]$

I think the grouped gemm interface needs to support heterogeneous $M$, $N$ and $K$, mirroring the PyTorch API.